### PR TITLE
Adds router package

### PR DESCRIPTION
**Pull request by milosgajdos**
This PR introduces `router` package which allows to implement micro network router. 

It provides two basic primitives:
* `Table` - routing table which allows to CRUD route operations
* `Router` -  maintains a routing table and advertises local services to "network" registry

The package provides basic e.g. default implementation of both `Router` and `Table` objects.

### CODE REVIEW

File: router/default_router.go [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+	"time"
+
+	"github.com/micro/go-micro/registry"
+	"github.com/olekukonko/tablewriter"
+)
+
+var (
+	// AdvertiseToNetworkTick defines how often in seconds do we scal the local registry
+	// to advertise the local services to the network registry
+	AdvertiseToNetworkTick = 5 * time.Second
```
Comment (asim) [inline router/default_router.go]: AdvertiseTick

File: router/default_router.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+	// AdvertiseNetworkTTL defines network registry TTL in seconds
+	// NOTE: this is a rather arbitrary picked value subject to change
+	AdvertiseNetworkTTL = 120 * time.Second
```
Comment (asim) [inline router/default_router.go]: AdvertiseTTL

File: router/options.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	NetworkAddress string
+	// LocalRegistry is router local registry
+	LocalRegistry registry.Registry
+	// NetworkRegistry is router remote registry
+	// NOTE: we need some abstraction on top of gossip.Registry
+	NetworkRegistry registry.Registry
+	// Table is routing table
+	Table Table
+	// RIB is Routing Information Base
+	RIB RIB
```
Comment (asim) [inline router/options.go]: Remove RIB as currently unused

File: router/options.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	// DefaultNetworkAddress is default network bind address
+	DefaultNetworkAddress = ":9094"
+)
+
+// Options are router options
+type Options struct {
+	// ID is router id
+	ID string
+	// Address is router micro service address
+	Address string
```
Comment (asim) [inline router/options.go:20]: Router has no address because it's not a service, its just a library

File: router/options.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	// GossipAddress is router gossip address
+	GossipAddress string
```
Comment (asim) [inline router/options.go]: Change gossip address to Address as no one knows its a gossip registry and the assumption is just this is the router's address to talk to other routers. We just happen to use gossip for it.

File: router/options.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	// NetworkAddress is micro network address
+	NetworkAddress string
```
Comment (asim) [inline router/options.go]: Change NetworkAddress to Advertise. This is the address which is advertised.

File: router/options.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	// LocalRegistry is router local registry
+	LocalRegistry registry.Registry
```
Comment (asim) [inline router/options.go]: LocalRegistry to Registry, its assumed to be the local registry

File: router/options.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	// NetworkRegistry is router remote registry
+	// NOTE: we need some abstraction on top of gossip.Registry
+	NetworkRegistry registry.Registry
```
Comment (asim) [inline router/options.go]: NetworkRegistry changed to Network.

File: router/query.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+	}
+}
+
+// QueryOption sets routing table query options
+type QueryOption func(*QueryOptions)
+
+// QueryOptions are routing table query options
+type QueryOptions struct {
+	// DestAddr is destination address
+	DestAddr string
```
Comment (asim) [inline router/query.go]: DestAddr to Destination

File: router/rib.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
+package router
+
+// RIB is Routing Information Base.
+// RIB is used to source the base routing table.
+type RIB interface {
```
Comment (asim) [inline router/rib.go]: RIB is being removed

File: router/route.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+var (
+	// DefaultLocalMetric is default route cost for local network
+	DefaultLocalMetric = 1
+	// DefaultNetworkMetric is default route cost for micro network
+	DefaultNetworkMetric = 10
+)
+
+// AddPolicy defines routing table addition policy
+type AddPolicy int
```
Comment (asim) [inline router/route.go]: Remove AddPolicy. Or change the Policy to be a function that takes a Route and allows the caller to define how to handle it e.g like Retry and Backoff Func in client https://github.com/micro/go-micro/blob/master/client/options.go#L51 then provide a default like AlwaysOverride

File: router/table_watcher.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+// WatchOption is used to define what routes to watch in the table
+type WatchOption func(*WatchOptions)
+
+// Watcher defines routing table watcher interface
+// Watcher returns updates to the routing table
+type Watcher interface {
+	// Next is a blocking call that returns watch result
+	Next() (*Event, error)
```
Comment (asim) [inline router/table_watcher.go:56]: Also add `Chan() (<-chan *Event, error)`. Been bitten by lack of Chan in registry watcher and we're going to add one.

File: router/table_watcher.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+type WatchOptions struct {
+	// Specify destination address to watch
+	DestAddr string
+	// Specify network to watch
+	Network string
+}
+
+// WatchDestAddr sets what destination to watch
+// Destination is usually microservice name
+func WatchDestAddr(a string) WatchOption {
```
Comment (asim) [inline router/table_watcher.go]: WatchDestination

File: router/table_watcher.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+	// Next is a blocking call that returns watch result
+	Next() (*Event, error)
+	// Stop stops watcher
+	Stop()
+}
+
+// WatchOptions are table watcher options
+type WatchOptions struct {
+	// Specify destination address to watch
+	DestAddr string
```
Comment (asim) [inline router/table_watcher.go]: Destination

File: router/query.go [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+	// NetworkAddress is network address
+	Network string
+	// Gateway is gateway address
+	Gateway Router
```
Comment (asim) [inline router/query.go]: Change to Router

File: router/options.go (hunk 4 above)
Comment (milosgajdos) [reply router/options.go:20]: We've agreed to "move" the `GossipAddress` into this field i.e. `Address` is the router's "gossip address"

Comment (asim) [reply router/options.go:20]: Agreed

**Pull request merged**